### PR TITLE
Fix #683 - User-side extension point

### LINKED ISSUE #683: [FEATURE] Allow customisation for eve callable outside of EVE namespace
See:

https://godbolt.org/z/KfeTfdxa3

for a POC

### PR DESCRIPTION
**Pull request by jfalcou**
USe ADL based tag dispatching system to allow users to define specific overload for EVE function without dealing with internals.

### CODE REVIEW

File: include/eve/detail/dispatcher.hpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace eve::detail
+{
+  template<typename Tag, typename Dummy = void> struct  callable_object_bis;
+
+  template <typename Tag, typename... Args>
+  concept tag_dispatchabe = requires(Tag tag, Args&&... args)
+  {
+    { tagged_dispatch(tag, std::forward<Args>(args)...) };
+  };
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/dispatcher.hpp:19]: Hmm. I'm actaully not sure why we need `do_tagged_dispatch`
Wouldn't concept guarntee adl lookup?

File: test/test_distribution.hpp [hunk 2]
```diff
@@ -161,8 +161,8 @@ namespace eve
     tests_integral_distribution() : tests_integral_distribution(valmin(as<T>()), valmax(as<T>())) { }
 
     tests_integral_distribution( T aa, T bb, int nbb = 300)
-      : a(eve::min(aa, bb)),
-        b(eve::max(aa, bb)),
+      : a(std::min(aa, bb)),
+        b(std::max(aa, bb)),
```
Comment (DenisYaroshevskiy) [inline test/test_distribution.hpp:165]: Did this break?

File: include/eve/module/real/proba/function/regular/generic/binomial_distribution.hpp [hunk 3]
```diff
@@ -465,13 +463,13 @@ namespace eve
...
+        return half(as<T>())*eve::log(twopie*d.n*d.p*d.q); // + �(1/n)
       else if constexpr (floating_value<U>)
-        return half(as<U>())*eve::log(twopie*d.n*U(0.25)); // + �(1/n)
+        return half(as<U>())*eve::log(twopie*d.n*U(0.25)); // + �(1/n)
       else if constexpr (floating_value<T>)
-        return half(as<T>())*eve::log(twopie*d.p*d.q); // + �(1/n)
+        return half(as<T>())*eve::log(twopie*d.p*d.q); // + �(1/n)
       else
-        return half(as<I>())*eve::log(twopie*I(0.25)); // + �(1/n)
+        return half(as<I>())*eve::log(twopie*I(0.25)); // + �(1/n)
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/proba/function/regular/generic/binomial_distribution.hpp:472]: Oh - can we not do unicode symbols?

**Review by DenisYaroshevskiy — COMMENTED** [review]
Small nit

**Review by DenisYaroshevskiy — APPROVED** [review]

File: include/eve/detail/overload.hpp [hunk 4]
```diff
@@ -14,59 +14,88 @@
...
+          check ( delay_t{}, ::eve::detail::types<tag_type>{}, std::forward<Arg>(d),               \
+                  std::forward<Args>(args)...                                                      \
+                );                                                                                 \
+        }                                                                                          \
+                                                                                                   \
+        if constexpr( tag_dispatchable<tag_type,Arg,Args...> )                                     \
+        {                                                                                          \
+          return dispatcher ( tag_type{}                                                           \
+                            , std::forward<Arg>(d), std::forward<Args>(args)...                    \
+                            );                                                                     \
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/overload.hpp:62]: I suspect  that you can just use `adl` here directly:

```
return tag_dispatch(tag_type{}, , std::forward<Arg>(d), std::forward<Args>(args)..);
```

Then dispatcher goes away. 

No need to change it, just saying.

File: include/eve/module/real/special/function/regular/generic/lgamma.hpp [hunk 5]
```diff
@@ -212,7 +212,7 @@ namespace eve::detail
           auto   ltza0 = is_ltz(a0);
           size_t nb    = eve::count_true(ltza0);
           T      r {0};
-          auto   other = [Logsqrt2pi](T x) {
+          auto   other = [Logsqrt2pi](T x)  -> T {
```
[the remark this answers is not in the document]
Comment (jfalcou) [reply include/eve/module/real/special/function/regular/generic/lgamma.hpp:215]: This is a remains of another fix. It happened that the issue comes from elsewhere, I'll need to clean this up

**Pull request merged**